python/cudf/cudf/core/reshape.py

```diff
@@ -63,6 +64,20 @@ def concat(objs, axis=0, ignore_index=False, sort=None):
         assert typs.issubset(allowed_typs)
         df = DataFrame()
 
+        # Step through objects to align indexes for pandas-like output
+        aligned = []
+        prev = list(_align_indices(objs[0], objs[1]))
+        for idx in range(len(objs) - 2):
+            aligned = []
+            for step in range(idx + 1):
+                aligned.append(_align_indices(prev[step], objs[idx + 2])[0])
+            aligned = aligned + list(
+                _align_indices(prev[idx + 1], objs[idx + 2])
+            )
+
+            prev = aligned
+        objs = aligned
```

Comment:
Can you explain why we need to nest the loops here? I imagine we need to loop over the objects once to calculate the new combined index, then a second loop to reindex all of the objects based on that new combined index, but we shouldn't need an inner loop, no?

Reply:
I'm going to double check my work here. I may have worked backwards and missed a step while breaking it down. I thought I could do it in a single loop initially and the indexes came up short. I suspect you are right though. Verifying.

Reply:
Ok. So the first looping I tried was a single loop going over the objects. dumping the aligned objects into a new list. I could not find a pattern that delivered the desired results as I described above.

```
platform linux -- Python 3.7.6, pytest-5.4.1, py-1.8.1, pluggy-0.13.0
rootdir: /home/mhooks/dev/rapids/cudf/python/cudf
plugins: forked-1.1.3, xdist-1.31.0, hypothesis-5.8.0
collecting ... 
collected 1 item                                                                                                                                                                                                           

cudf/tests/test_concat.py 
got
                      a             b             c
ind                                                
-0.771553          null  -0.308155346          null
-0.515582          null          null   -0.03390487
-0.306220  -0.623033436          null  -0.667679473
 0.161977   0.175154322  -0.843013493          null
 0.213204  -0.783469252          null  -0.747134441
 0.945599          null  -0.385942423          null
expect
                  a         b         c
ind                                    
-0.771553       NaN -0.308155       NaN
-0.606135       NaN       NaN -0.033905
-0.569325       NaN       NaN -0.667679
-0.515582 -0.623033       NaN       NaN
-0.306220  0.175154       NaN       NaN
 0.161977 -0.783469       NaN       NaN
 0.213204       NaN -0.843013       NaN
 0.491941       NaN       NaN -0.747134
 0.945599       NaN -0.385942       NaN
```

I found that to get the correct index I needed to take a chunk of two, then align the next two with those two, and so on, and so on. Otherwise for a concat of 3 dataframes I would end up with the above 6x3. I may be missing something simple about how to align 3+ indexes after I get them in one pass when the operation I'm using only aligns in chunks of two. I was just sitting here trying to get to what you're talking about and I just don't know how to get the index then align and order without two loops.

Here is the exploded logic I'm working from, let me know if this gives you any insight into what I'm missing. Links to existing `cudf` docs with any examples are sufficient.

```
prev = _align_indices(objs[0], objs[1])
        aligned = []
        
        aligned.append(_align_indices(prev[0], objs[2])[0])
        aligned = aligned + list(_align_indices(prev[1], objs[2]))
        
        prev = aligned
        aligned = []
        
        aligned.append(_align_indices(prev[0], objs[3])[0])
        aligned.append(_align_indices(prev[1], objs[3])[0])        
        aligned = aligned + list(_align_indices(prev[2], objs[3]))
        
        prev = aligned
        aligned = []
        
        aligned.append(_align_indices(prev[0], objs[4])[0])
        aligned.append(_align_indices(prev[1], objs[4])[0])
        aligned.append(_align_indices(prev[2], objs[4])[0])        
        aligned = aligned + list(_align_indices(prev[3], objs[4]))
```